### PR TITLE
feat: return reason for failed authorisation

### PR DESCRIPTION
**Pull request by nichonien**

### CODE REVIEW

**Review by JGiter — CHANGES_REQUESTED** [review]
If possible please replace error strings with error classes from vc-verification

File: lib/LoginStrategy.ts [hunk 1]
```diff
@@ -287,19 +278,44 @@ export class LoginStrategy extends BaseStrategy {
...
+            (verifiedRole) => verifiedRole.namespace === role
+          );
+          if (!acceptedRole) {
+            invalidRoles.push({
+              name: role,
+              namespace: role,
+              error: 'Role credential not found for the user.',
+            });
+            userValidationStatus = false;
+          } else if (acceptedRole.error) {
```
[the remark this answers is not in the document]
Comment (nichonien) [reply lib/LoginStrategy.ts:298]: I think the naming of `acceptedRole` is wrong, it is not related to `acceptedRoles` but it is a declaration to process intermediate values. I will change it.

File: lib/ClaimVerifier.ts [hunk 2]
```diff
@@ -48,6 +49,7 @@ export class ClaimVerifier {
   private async verifyRole(credential: Required<RoleEIP191JWT>): Promise<{
     name: IRoleDefinitionV2['roleName'];
     namespace: RoleEIP191JWT['payload']['claimData']['claimType'];
+    error: string;
```
[the remark this answers is not in the document]
Comment (nichonien) [reply lib/ClaimVerifier.ts:52]: Yeah, it eases the validation based on null. Like this https://github.com/energywebfoundation/passport-did-auth/blob/2f0d23d8d7bb3754cdad6827cd2d8c96d6a8a2d6/lib/LoginStrategy.ts#L298

File: lib/LoginStrategy.ts [hunk 3]
```diff
@@ -422,4 +401,41 @@ export class LoginStrategy extends BaseStrategy {
 
     return true;
   }
+
+  /**
+   * Filters verified roles into valid and invalid roles and sets validation status
+   * @param uniqueRoles verified roles
+   * @returns
+   */
+  filterUserRoleStatus(uniqueRoles: RoleStatus[]): {
```
Comment (JGiter) [inline lib/LoginStrategy.ts:410]: Is this method needs to be exposed?

File: lib/LoginStrategy.ts [hunk 4]
```diff
@@ -422,4 +401,41 @@ export class LoginStrategy extends BaseStrategy {
 
     return true;
   }
+
+  /**
+   * Filters verified roles into valid and invalid roles and sets validation status
```
Comment (JGiter) [inline lib/LoginStrategy.ts:406]: I beleive that `invalid` isn't appropriate in this case. May be `notSaticfiedRoles`. Also `valid` isn't needed, because user most often need to know which requirements he yet need to satisfy

File: lib/ClaimVerifier.ts [hunk 5]
```diff
@@ -32,6 +33,7 @@ export class ClaimVerifier {
     const filteredRoles = roles.filter(Boolean) as {
       name: IRoleDefinitionV2['roleName'];
       namespace: RoleEIP191JWT['payload']['claimData']['claimType'];
+      error: string;
```
Comment (JGiter) [inline lib/ClaimVerifier.ts:36]: Its still looks odd to me that `getVerifiedRoles` returns not verified role

**Review by JGiter — CHANGES_REQUESTED** [review]

File: lib/LoginStrategy.ts (hunk 3 above)
Comment (nichonien) [reply lib/LoginStrategy.ts:410]: No, will make it private .

File: lib/LoginStrategy.ts (hunk 4 above)
Comment (nichonien) [reply lib/LoginStrategy.ts:406]: Valid credentials are needed, since we display the verified roles once authorised in `ram-client-examples`. PDA has always returned verified roles.

File: lib/ClaimVerifier.ts (hunk 5 above)
Comment (nichonien) [reply lib/ClaimVerifier.ts:36]: 1. To me verification is about ensuring the data came from the right channel and its integrity is still persisted. Validity is solely based on business use case and it is next step after verification. For example - I was issued driving license which was valid for 10 years. Even after expiration I can use it to prove my age at a Bar but I cannot drive without renewal.
2. `getVerifiedRoles` is solely responsible for verification and if one wants to also know the invalid roles (expired or revoked) then it will come from this call itself. 

Maybe we can segregate the valid and revoked / expired role here itself so that there is no confusion.

File: lib/LoginStrategy.ts (hunk 4 above)
Comment (JGiter) [reply lib/LoginStrategy.ts:409]: Ok, thanks. Just to clarify, is it assumed that `validate` returns only those user roles which present in accepted?

Comment (nichonien) [reply lib/LoginStrategy.ts:409]: It depends on another parameter `includeAllRoles` - https://github.com/energywebfoundation/passport-did-auth/blob/6028f44f026b615e93db5095b4c530c7d1d5c1b4/lib/LoginStrategy.ts#L66.

So, if this is true it will return all verified roles and if not then only the accepted roles.

Comment (JGiter) [reply lib/LoginStrategy.ts:409]: But isn't `filterUserRoleStatus` rerurns only accepted roles regardless of `includeAllRoles`?

Comment (JGiter) [reply lib/LoginStrategy.ts:409]: I think we just shouldn't call `filterUsrRoleStatus`  when `includeAllRoles`

Comment (nichonien) [reply lib/LoginStrategy.ts:409]: @JGiter Thinking more on the above, what should we do of revoked or expired roles if the `includeAllRoles` is true? Should it be separated in a different list?

Comment (jrhender) [reply lib/LoginStrategy.ts:418]: @JGiter @nichonien where is `filterUserRoleStatus`? I don't see it in this code

> Thinking more on the above, what should we do of revoked or expired roles if the includeAllRoles is true? Should it be separated in a different list?

@nichonien does your question about still stand? To me, they can just be roles in the `userRoles` array that are marked as `revoked` or `expired`

File: lib/LoginStrategy.ts [hunk 6]
```diff
@@ -406,4 +413,38 @@ export class LoginStrategy extends BaseStrategy {
...
+            name: role,
+            namespace: role,
+            error: 'Role credential not found for the user.',
+          });
+          authorisationStatus = false;
+        } else if (userRole.error) {
+          authorisationStatus = false;
+        }
+      });
+    }
```
Comment (jrhender) [inline lib/LoginStrategy.ts:444]: @nichonien Is this is a change in the behaviour from before?
I think the previous code (below) passed if **any** of the roles from `acceptedRoles` where verified and present. I think the new code requires that  **all** accepted roles are present?
If this is indeed a change, was this change intended?
```
this.acceptedRoles.size > 0 &&
!uniqueRoles.some(({ namespace }) => {
          return this.acceptedRoles.has(namespace);
})
```

**Review by jrhender — CHANGES_REQUESTED** [review]
The change in behaviour in `validateAcceptedRoles` from any to all is a potential issue

File: lib/LoginStrategy.ts (hunk 4 above)
Comment (nichonien) [reply lib/LoginStrategy.ts:418]: > @JGiter @nichonien where is filterUserRoleStatus? I don't see it in this code

I renamed it since we decided to return just one array of roles.

File: lib/LoginStrategy.ts (hunk 6 above)
Comment (nichonien) [reply lib/LoginStrategy.ts:444]: @jrhender No, this is a mistake. I will correct it.

Comment (jrhender) [reply lib/LoginStrategy.ts:444]: Ok, great. Maybe we can add a test for this case (should be authorized when only one of the acceptedRoles is verified and other is not verified) either in this PR or a separate one?

Comment (nichonien) [reply lib/LoginStrategy.ts]: Adding a test is not possible, since mocking `ClaimVerifier` is not feasible. Our earlier discussion - https://github.com/energywebfoundation/passport-did-auth/pull/331#discussion_r985457467

File: lib/ClaimVerifier.ts [hunk 7]
```diff
@@ -32,6 +34,7 @@ export class ClaimVerifier {
     const filteredRoles = roles.filter(Boolean) as {
```
Comment (JGiter) [inline lib/ClaimVerifier.ts:34]: Why to cast to `{name; namespace; status}`?

**Review by JGiter — APPROVED** [review]

File: lib/ClaimVerifier.ts (hunk 7 above)
Comment (nichonien) [reply lib/ClaimVerifier.ts:34]: @JGiter I tried removing this code and it results in some abnormalities in returned `verifiedRoles`. Maybe I give it some more time and remove this in a separate PR? So that the UI in `iam-client-examples` could pick this informative rejection changes ASAP. @jrhender what are your thoughts?

File: lib/ClaimVerifier.ts [hunk 8]
```diff
@@ -63,8 +67,16 @@ export class ClaimVerifier {
...
-      return null;
+    const roleStatus = await this.checkRoleStatus(credential.payload);
+    if (!roleStatus.verified) {
+      return {
+        name: role.roleName,
+        namespace: credentialClaimData.claimType,
+        status:
+          roleStatus.error === CredentialRevoked
+            ? RoleCredentialStatus.REVOKED
+            : RoleCredentialStatus.EXPIRED,
```
Comment (jrhender) [inline lib/ClaimVerifier.ts:78]: The code as-is is a bit brittle because if a new error is added but this code isn't changed, then `RoleCredentialStatus.EXPIRED` would be returned incorrectly. What about this change below?
```suggestion
      const errorMessageMap = {
        CredentialExpired: RoleCredentialStatus.EXPIRED,
        CredentialRevoked: RoleCredentialStatus.REVOKED
      }
      return {
        name: role.roleName,
        namespace: credentialClaimData.claimType,
        status: errorMessageMap[roleStatus.error] ?? RoleCredentialStatus.UNKNOWN
```

File: lib/ClaimVerifier.ts (hunk 7 above)
Comment (jrhender) [reply lib/ClaimVerifier.ts:34]: > Maybe I give it some more time and remove this in a separate PR

I agree, better to do it in a separate PR

File: lib/ClaimVerifier.ts (hunk 8 above)
Comment (nichonien) [reply lib/ClaimVerifier.ts:78]: @jrhender I agree with you, but then if a new `error` is added then we would need to update error mapping and add error to `RoleCredentialStatus` as well to make it meaningful. So the code would be needed to update anyhow.

**Review by jrhender — APPROVED** [review]
Looks good, thanks!
